nuxeo-runtime/nuxeo-runtime-mongodb/src/main/java/org/nuxeo/runtime/mongodb/MongoDBComponent.java

```diff
@@ -103,8 +111,78 @@ public MongoDatabase getDatabase(String id) {
     public Iterable<MongoDatabase> getDatabases() {
         return () -> clients.entrySet().stream().map(e -> {
             MongoDBConnectionConfig c = getDescriptor(XP_CONNECTION, e.getKey());
-            return MongoDBConnectionHelper.getDatabase(e.getValue(), c.dbname);
+            MongoDatabase db = MongoDBConnectionHelper.getDatabase(e.getValue(), c.dbname);
+            MongoDBCountHelper.versions.putIfAbsent(db, fetchVersion(db));
+            return db;
         }).iterator();
     }
 
+    /**
+     * @return the simplified version of the database server. This string will take the format <major>.<minor>.<patch>
+     *         in the case of a release, but development builds may contain additional information.
+     * @since 11.1
+     */
+    protected String fetchVersion(MongoDatabase db) {
+        Document result = db.runCommand(new Document("buildInfo", 1));
+        return (String) result.get("version");
+    }
+
+    /**
+     * This helper is intended only for Nuxeo 10.X to ensure compatibility with MongoDB [3.4,3.8[
+     *
+     * @since 11.1
+     */
+    public static class MongoDBCountHelper {
+
+        private MongoDBCountHelper() {
+            // Empty
+        }
+
+        protected static final Map<MongoDatabase, String> versions = new HashMap<>();
```

Comment:
I don't understand why the key cannot be the `databaseName` (which we can get easily from the collection, and therefore avoid passing the database in all calls to `countDocuments`). I don't understand the comment in NXP-27953 either.

Are you afraid of two different database ids connecting to completely different databases but that have the same database name? If that's the case then we can ignore that issue and assume that all connected databases MUST have the same version, so actually we can simplify the code: record only one version instead of a map, and make `recordVersion` throw an error if the db version is different than the previously recorded one (if it exists).

Reply:
If all databases MUST have the same version, then it's a ready made path as you said with no need for a map and extra args and all. In fact we wouldn't even need to access the collection's database name. 
But what I understood is that you can have several mongos with different versions. I think we need to sync everybody on the scenarios to support or not ?